src/hooks/users.ts

```diff
@@ -1,15 +1,97 @@
+import { universeUsersSelector, usersByIdSelector } from "utils/selectors";
 import { WithId } from "utils/id";
 
 import { useSelector } from "hooks/useSelector";
 import { useUserLastSeenThreshold } from "hooks/useUserLastSeenThreshold";
 import { useVenueId } from "hooks/useVenueId";
 import { useSparkleFirestoreConnect } from "hooks/useSparkleFirestoreConnect";
+import { useConnectRelatedVenues } from "hooks/useConnectRelatedVenues";
 
 import { User } from "types/User";
+import { useConnectCurrentVenueNG } from "./useConnectCurrentVenueNG";
+
+export const useRecentPartyUsers = () => [];
+// export const useRecentPartyUsers = (venueName: string): WithId<User>[] => {
+//   const venueId = useVenueId();
+//   const partygoers = useSelector(partygoersSelector) ?? [];
+
+//   const { parentVenue, relatedVenues } = useConnectRelatedVenues({
+//     venueId,
+//   });
+
+//   const lastSeenThreshold = useUserLastSeenThreshold();
+
+//   const parentVenueName = parentVenue?.name ?? "";
+
+//   const filterPartygoersByHierarchy = useCallback(
+//     (partygoer: WithId<User>) => {
+//       return relatedVenues?.map((venue) => {
+//         return (
+//           partygoer?.lastSeenIn?.[parentVenueName || venueName || venue.name] >
+//           lastSeenThreshold
+//         );
+//       });
+//     },
+//     [lastSeenThreshold, parentVenueName, relatedVenues, venueName]
+//   );
+
+//   return useMemo(() => partygoers.filter(filterPartygoersByHierarchy), [
+//     partygoers,
+//     filterPartygoersByHierarchy,
+//   ]);
+// };
+
+// NOTE: Heuristic approach. Way faster and efficient.
+// Is based on assumption that there is only one parent venue and it is an entry point for any user
+export const useConnectUniverseUsers = () => {
+  const venueId = useVenueId();
+
+  const { currentVenue, isCurrentVenueLoaded } = useConnectCurrentVenueNG(
+    venueId
+  );
+
+  useSparkleFirestoreConnect(() => {
+    if (!isCurrentVenueLoaded || !currentVenue) return [];
+
+    const rawVenueId = currentVenue.parentId || currentVenue.name;
+    const venueId = rawVenueId.toLowerCase();
+    return [
+      {
+        collection: "users",
+        where: ["enteredVenueIds", "array-contains", venueId],
+        storeAs: "universeUsers",
+      },
+    ];
+  });
+};
+
+// NOTE: Logically correct version of pulling users
+export const useConnectWorldUsers_V2 = () => {
+  const venueId = useVenueId();
+
+  const { relatedVenues, isRelatedVenuesLoaded } = useConnectRelatedVenues({
+    venueId,
+  });
+
+  useFirestoreConnect(() => {
+    if (!isRelatedVenuesLoaded) return [];
+
+    const worldUsersVenueNames = relatedVenues.map((venue) =>
+      venue.name.toLowerCase()
+    );
+
+    console.log({ worldUsersVenueNames });
+
+    return worldUsersVenueNames.map((venueName) => ({
+      collection: "users",
+      where: ["enteredVenueIds", "array-contains", venueName],
+      storeAs: "universeUsers",
+    }));
```

Comment:
It's possible to setup our own manual firestore 'watch queries' without using the `useFirestoreConnect` hook:

- http://react-redux-firebase.com/docs/firestore.html#manual

That would allow us to avoid the `storeAs` naming collision/hacks/etc.

We would probably want to wrap this into some kind of custom hook that handles the `setListener` / `unsetListener` in a `useEffect` (so that it properly cleans up after itself).

Based on our 'main root venue', we could find any venues that have it as their `parentId`, then for each of those 'first level children' search for any venues that have them as a `parentId`, etc until we have 'walked the full tree' of parent -> children -> sub-children -> etc.

If we didn't care about this being able to 'live detect' new venues that get created (and were happy for a browser refresh to be required for that) then we could probably get away with running this once on app load, even if that is only to collect a list of all related venue ids.

This is the kind of feature I was thinking we could implement 'nicer' on the backend, potentially using firebase hooks that would automatically run this code and then save it somewhere in the DB as a single 'lookup table' reference point. eg. Having an onCreated/onUpdated hook that detects new/edited venues, then 'walks the tree' to find related venues, and saves that result somewhere in the DB.